tests/testshell/mod.rs

```diff
@@ -0,0 +1,210 @@
+extern crate pty;
+extern crate vte;
+
+use std::os::unix::process::CommandExt;
+use std::process::Command;
+use std::io::Read;
+use std::io::Write;
+use std::time::Duration;
+use std::sync::mpsc;
+use std::thread;
+
+pub struct TestShell {
+    // fork is here for lifetime reasons; can't drop it until the pty is done
+    fork: pty::fork::Fork,
+    pty: pty::fork::Master,
+    output: mpsc::Receiver<String>,
+    eof: mpsc::Receiver<()>,
+}
+
+// VTEData is to handle lines after the mess of vte terminal stuff.
+// It keeps track of newlines and such
+#[derive(PartialEq, Clone, Debug)]
+struct VTEData {
+    current_line_cursor: usize,
+    pub current_line: String,
+    pub scrollback: Vec<String>,
+}
+
+impl VTEData {
+    fn new() -> Self {
+        VTEData {
+            current_line_cursor: 0,
+            current_line: String::new(),
+            scrollback: Vec::new(),
+        }
+    }
+}
+
+impl vte::Perform for VTEData {
+    fn print(&mut self, c: char) {
+        self.current_line.truncate(self.current_line_cursor);
+        self.current_line_cursor += 1;
+        self.current_line.push(c);
+    }
+
+    fn execute(&mut self, byte: u8) {
+        if byte as char == '\n' {
```

Comment:
Couldn't (shouldn't) this be pattern matching?

Reply:
Could be, though the fact that `'\n'` exists and `'\b'` doesn't led me to using char / byte depending on which I'm looking for.
I'll see how switching it to a match looks ...